metaflow/plugins/pypi/pip.py

```diff
@@ -72,20 +74,117 @@ def solve(self, id_, packages, python, platform):
                 # *(chain.from_iterable(product(["--implementations"], set(implementations)))),
             ]
             for package, version in packages.items():
-                if version.startswith(("<", ">", "!", "~")):
+                if version.startswith(("<", ">", "!", "~", "@")) or version == "":
                     cmd.append(f"{package}{version}")
                 else:
                     cmd.append(f"{package}=={version}")
             self._call(prefix, cmd)
+
+            def _format_item(item):
+                dl_info = item["download_info"]
+                res = {k: v for k, v in dl_info.items() if k in ["url"]}
+                # If source url is not a wheel, we need to build the target. Add a build flag.
+                res["require_build"] = not res["url"].endswith(".whl")
+
+                # reconstruct the VCS url and pin to current commit_id
+                # so using @branch as a version acts somewhat as expected.
+                vcs_info = dl_info.get("vcs_info")
+                if vcs_info:
+                    res["url"] = "{vcs}+{url}@{commit_id}".format(**vcs_info, **res)
+                    # used to deduplicate the storage location in case wheel does not build with enough unique identifiers.
+                    res["hash"] = vcs_info["commit_id"]
+                return res
+
             with open(report, mode="r", encoding="utf-8") as f:
-                return [
-                    {k: v for k, v in item["download_info"].items() if k in ["url"]}
-                    for item in json.load(f)["install"]
+                return [_format_item(item) for item in json.load(f)["install"]]
+
+    def build_wheels(self, id_, packages, python, platform):
+        prefix = self.micromamba.path_to_environment(id_)
+        build_metadata_file = BUILD_METADATA_FILE.format(prefix=prefix)
+        # skip build if it has been tried already.
+        if os.path.isfile(build_metadata_file):
+            return
+        metadata = {}
+        with ThreadPoolExecutor() as executor:
+            results = list(
+                executor.map(
+                    lambda x: self._build_wheel(*x, prefix), enumerate(packages)
+                )
+            )
+
+            def _grab_wheel_from_path(path):
+                wheels = [
+                    f
+                    for f in os.listdir(path)
+                    if os.path.isfile(os.path.join(path, f)) and f.endswith(".whl")
                 ]
+                if len(wheels) != 1:
+                    raise PipException(
+                        "Incorrect number of wheels found in path %s after building: %s"
+                        % (path, wheels)
+                    )
+                return wheels[0]
+
+            def _verify_wheel_tags_supported(wheel_name):
+                target_tags = pip_tags(python, platform)
+                built_tags = tags_from_wheel_name(wheel_name)
+                if built_tags not in target_tags:
```

Comment:
I don't think this is quite correct. You probably want a set intersection and if empty, then error out.

Reply:
Great catch! Fixed this and added some unit tests for validation. Coverage is not extensive but should at least handle obvious cross-platform deployment breaks (e.g. `osx-arm64` -> `linux-64`)